dolphinscheduler-task-plugin/dolphinscheduler-task-api/src/main/java/org/apache/dolphinscheduler/plugin/task/api/am/KubernetesApplicationManager.java

```diff
@@ -0,0 +1,201 @@
+        }
+        return watchList;
+    }
+
+    /**
+     * create client or get from cache map
+     *
+     * @param kubernetesApplicationManagerContext
+     * @return
+     */
+    private KubernetesClient getClient(KubernetesApplicationManagerContext kubernetesApplicationManagerContext) {
+        K8sTaskExecutionContext k8sTaskExecutionContext =
+                kubernetesApplicationManagerContext.getK8sTaskExecutionContext();
+        Config config = Config.fromKubeconfig(k8sTaskExecutionContext.getConfigYaml());
+        return cacheClientMap.computeIfAbsent(kubernetesApplicationManagerContext.getLabelValue(),
+                key -> new DefaultKubernetesClient(config));
+    }
+
+    public void removeCache(String cacheKey) {
+        KubernetesClient client = cacheClientMap.get(cacheKey);
+        if (Objects.nonNull(client)) {
+            client.close();
+        }
+        cacheClientMap.remove(cacheKey);
+    }
+
+    /**
+     * get application execution status
+     *
+     * @param kubernetesApplicationManagerContext
+     * @return TaskExecutionStatus  SUCCESS / FAILURE
+     * @throws TaskException
+     */
+    public TaskExecutionStatus getApplicationStatus(KubernetesApplicationManagerContext kubernetesApplicationManagerContext) throws TaskException {
+        return getApplicationStatus(kubernetesApplicationManagerContext, null);
+    }
+
+    /**
+     * get application (driver pod) status
+     *
+     * @param kubernetesApplicationManagerContext
+     * @param watchList
+     * @return
+     * @throws TaskException
+     */
+    private TaskExecutionStatus getApplicationStatus(KubernetesApplicationManagerContext kubernetesApplicationManagerContext,
+                                                     FilterWatchListDeletable<Pod, PodList> watchList) throws TaskException {
+        String phase;
+        try {
+            if (Objects.isNull(watchList)) {
+                watchList = getDriverPod(kubernetesApplicationManagerContext);
+            }
+            List<Pod> driverPod = watchList.list().getItems();
+            if (!driverPod.isEmpty()) {
+                // cluster mode
+                Pod driver = driverPod.get(0);
+                phase = driver.getStatus().getPhase();
+            } else {
+                // client mode
+                phase = FINISH;
+            }
+        } catch (Exception e) {
+            throw new TaskException("Failed to get Kubernetes application status", e);
+        }
+
+        return phase.equals(FAILED) || phase.equals(UNKNOWN) ? TaskExecutionStatus.FAILURE
+                : TaskExecutionStatus.SUCCESS;
+    }
+
+    /**
+     * collect pod's log
+     *
+     * @param kubernetesApplicationManagerContext
+     * @return
+     */
+    public String collectPodLog(KubernetesApplicationManagerContext kubernetesApplicationManagerContext) {
+        try (KubernetesClient client = getClient(kubernetesApplicationManagerContext)) {
```

Comment:
Here the `client` is closed automatically, but in `killApplication` (line 192), the `client` will be closed again, (`killApplication` --> `removeCache`)

Reply:
It should be closed in `killApplication`, cuz when master do failover, `collectPodLog` will not be called, i forgot to remove it, thx!